cmd/snap/cmd_run.go

```
@@ -227,6 +230,143 @@ func isReexeced() bool {
+	// but it is better to convert the uid from the stat result to a
+	// string than a string into a number.
+	if fmt.Sprintf("%d", sys.(*syscall.Stat_t).Uid) != u.Uid {
+		return "", fmt.Errorf(i18n.G("Xauthority file isn't owned by the current user %s"), u.Uid)
+	}
+
+	targetPath := filepath.Join(baseTargetDir, ".Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// otherwise we can continue using the existing Xauthority file.
+	// This is ok to do here because we aren't trying to protect against
+	// the user changing the Xauthority file in XDG_RUNTIME_DIR outside
+	// of snapd.
+	var fout *os.File
+	if osutil.FileExists(targetPath) {
+		if fout, err = os.Open(targetPath); err != nil {
+			return "", err
+		}
+		defer fout.Close()
+		if osutil.FileStreamsEqual(fin, fout) {
+			return targetPath, nil
+		}
+	}
+
+	// Ensure we're validating the Xauthority file from the beginning
+	if _, err := fin.Seek(int64(os.SEEK_SET), 0); err != nil {
+		return "", err
+	}
+
+	// To guard against setting XAUTHORITY to non-xauth files, check
+	// that we have a valid Xauthority. Specifically, the file must be
+	// parseable as an Xauthority file and not be empty.
+	if err := x11.ValidateXauthorityFromFile(fin); err != nil {
+		logger.Noticef("WARNING: invalid Xauthority file: %s", err)
+		return "", nil
+	}
+
+	if osutil.FileExists(targetPath) {
+		err := os.Remove(targetPath)
+		if err != nil {
+			logger.Noticef("WARNING: failed to remove existing file %s", targetPath)
+		}
+	}
+
+	// Read data from the beginning of the file
+	if _, err = fin.Seek(int64(os.SEEK_SET), 0); err != nil {
+		return "", err
+	}
+
+	fout, err = os.OpenFile(targetPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	defer fout.Close()
+	if err != nil {
+		return "", err
+	}
+
+	// Read and write validated Xauthority file to its right location
+	if _, err = io.Copy(fout, fin); err != nil {
+		err := os.Remove(targetPath)
```

This could be a single line `if err := os.Remove(...); err != nill`. Also maybe slightly confusing, if the copy fails and the remove also fails we swallow one of the errors. Maybe logger.Notice about the remove failure? Or going back to your original code and simply ignoring the error. Sorry for complicating things.